src/features/organizations/components/OrganizationSwitcher/useOrgSwitcher.ts

```diff
@@ -0,0 +1,99 @@
+import Fuse from 'fuse.js';
+import { RootState } from 'core/store';
+import { TreeItemData } from 'features/organizations/types';
+import useLocalStorage from 'zui/hooks/useLocalStorage';
+import { useMemo } from 'react';
+import { useSelector } from 'react-redux';
+
+function useOrgSwitcher(orgId: number, searchString: string) {
+  function notEmpty<TValue>(value: TValue | null | undefined): value is TValue {
+    return value !== null && value !== undefined;
+  }
+
+  const treeDataList = useSelector(
+    (state: RootState) => state.organizations.treeDataList
+  );
+  const orgData = treeDataList.items.map((item) => item.data).filter(notEmpty);
+
+  const flatOrgData = makeFlatOrgData(orgData);
+
+  const [recentOrganizationIds, setRecentOrganizationIds] = useLocalStorage(
+    'recentOrganizationIds',
+    [] as number[]
+  );
+
+  function makeFlatOrgData(orgData: TreeItemData[]): TreeItemData[] {
+    let children = [] as TreeItemData[];
+    const flatOrgData = orgData.map((org) => {
+      if (org.children && org.children.length) {
+        children = [...children, ...org.children];
+      }
+      return org;
+    });
+
+    return flatOrgData.concat(
+      children.length ? makeFlatOrgData(children) : children
+    );
+  }
+
+  function onSwitchOrg() {
+    setRecentOrganizationIds([
+      orgId,
+      ...recentOrganizationIds.filter((id) => id != orgId),
+    ]);
+  }
+
+  const recentOrgs = recentOrganizationIds
+    .map((id) => flatOrgData.find((org) => org.id === id))
+    .filter((org) => org?.id != orgId);
+
+  const filteredRecentOrgs = useMemo(() => {
+    const fuse = new Fuse(recentOrgs, {
+      keys: ['title'],
+      threshold: 0.4,
+    });
+
+    return searchString
+      ? fuse.search(searchString).map((fuseResult) => fuseResult.item)
+      : flatOrgData;
+  }, [searchString]);
+
+  const filteredAllOrgs = useMemo(() => {
+    const fuse = new Fuse(flatOrgData, {
+      keys: ['title'],
+      threshold: 0.4,
+    });
+
+    return searchString
+      ? fuse.search(searchString).map((fuseResult) => fuseResult.item)
+      : flatOrgData;
+  }, [searchString]);
```

Comment:
Using memo here means that the search results are memoized as long as the search string doesn't change. That helps against searches happening unnecessarily after some state other than the search string changes, which is great.

However, two things are happening here. The first is that a `Fuse` instance is created, which builds the index. Building the index takes significant time. Notice how it doesn't take the search string as a parameter. That's because the index depends only on the data, and can be kept between searches.

So this could actually be optimized further by splitting this up into two memoized operations.

```suggestion
  const fuse = useMemo(() => {
    return new Fuse(flatOrgData, {
      keys: ['title'],
      threshold: 0.4,
    });
  }, [flatOrgData]);
  
  const filteredAllOrgs = useMemo(() => {
    return searchString
      ? fuse.search(searchString).map((fuseResult) => fuseResult.item)
      : flatOrgData;
  }, [searchString]);
```